SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizer.cc

```diff
@@ -269,26 +279,91 @@ namespace cms {
+
+            bool checkTwoSimHits = false;
+            if (channelPrevious2 == loopNewClass->channel() && hitFirstOne2 != loopNewClass->hitIndex()) {
+              // case of one Digi associated to a second SimHit
+              // std::cout << "     --> identification of the same Digi associated to anothe SimHit ! " << std::endl;
+              checkTwoSimHits = true;
+            } else {
+              channelPrevious2 = loopNewClass->channel();
+              hitFirstOne2 = loopNewClass->hitIndex();
+            }
+
+            bool checkInTheList = false;
+            if (!checkTwoSimHits) {
+              std::vector<PixelSimHitExtraInfo>::iterator loopTempSH;
+              for (loopTempSH = tempSHcollector.begin(); loopTempSH != tempSHcollector.end(); ++loopTempSH) {
+                if (loopNewClass->hitIndex() == loopTempSH->hitIndex()) {
+                  checkInTheList = true;
+                  loopTempSH->addDigiInfo(loopNewClass->channel());
+                }
+              }
+              if (!checkInTheList) {
+                PixelSimHitExtraInfo newSHEntry(loopNewClass->hitIndex(),
+                                                loopNewClass->entryPoint(),
+                                                loopNewClass->exitPoint(),
+                                                loopNewClass->channel());
+                tempSHcollector.push_back(newSHEntry);
+              }
+            }
+          }
+        }
+
+        if (applyLateReweighting_) {
+          // if applyLateReweighting_  is true, the charge reweighting has to be applied on top of the digis
+          _pixeldigialgo->lateSignalReweight(
+              dynamic_cast<const PixelGeomDetUnit*>(iu), collector.data, tempSHcollector.data, tTopo, randomEngine_);
+        }
 
-        _pixeldigialgo->digitize(
-            dynamic_cast<const PixelGeomDetUnit*>(iu), collector.data, linkcollector.data, tTopo, randomEngine_);
         if (!collector.data.empty()) {
           theDigiVector.push_back(std::move(collector));
         }
         if (!linkcollector.data.empty()) {
           theDigiLinkVector.push_back(std::move(linkcollector));
         }
+        if (!tempSHcollector.data.empty()) {
+          theExtraSimHitInfoVector.push_back(std::move(tempSHcollector));
+        }
       }
     }
+    _pixeldigialgo->resetSimHitMaps();
 
     // Step C: create collection with the cache vector of DetSet
     std::unique_ptr<edm::DetSetVector<PixelDigi> > output(new edm::DetSetVector<PixelDigi>(theDigiVector));
     std::unique_ptr<edm::DetSetVector<PixelDigiSimLink> > outputlink(
         new edm::DetSetVector<PixelDigiSimLink>(theDigiLinkVector));
+    std::unique_ptr<edm::DetSetVector<PixelSimHitExtraInfo> > outputExtraSim(
+        new edm::DetSetVector<PixelSimHitExtraInfo>(theExtraSimHitInfoVector));
 
     // Step D: write output to file
     iEvent.put(std::move(output));
     iEvent.put(std::move(outputlink));
+    iEvent.put(std::move(outputExtraSim));
```

Comment:
Same here, would it make sense to have this under `if(applyLateReweighting_)`  to avoid putting an empty collection in the event .